SalesforceCommerceCloud/pwa-kit
# Support storing authentication tokens in cookie

# Description



Today, in retail react app, the authentication tokens such as `usid`, `enc-user-id`, SCAPI bearer token are stored in localstorage. We'd like to move the tokens from localstorage to cookies to support Hybrid Deployment.

### What is hybrid deployment?

Hybrid deployment allow projects to launch new PWA storefronts for a partial of the sites, alongside with existing SFRA site. A typical hybrid build contains the top of the funnel pages like homepage, PLP, PDP, search pages in the PWA and have checkout pages from the SFRA site. This minimize the launch risks and enable teams to launch PWA site in incremental phases.

### Why cookies?

To enable shoppers to navigate between the sites seamlessly, we need a mechanism to store SLAS refresh token and OCAPI session tokens in a single place for both SFRA and PWA. With the recent launch of [SLAS cartridge](https://github.com/SalesforceCommerceCloud/plugin_slas), the SLAS refresh tokens `cc-nx` are stored in cookies, as well as the OCAPI session token `dwsid`.

### Changes

- move the existing tokens `usid`, `enc-user-id`, `oid` and the SCAPI bearer token `token` from localstorage to cookies
- rename SLAS refresh token `refresh-token` to `cc-nx` (registered user) and `cc-nx-g` (guest user), and store them as cookies
- Upon shopper login, bridge the OCAPI session by making a POST request to `/sessions` endpoint

## Todo
- [x] `js-cookie` 3PP approval: https://gus.lightning.force.com/lightning/r/ADM_Third_Party_Software__c/a0qEE00000023mbYAA/view
- [x] double check with [plugin_slas](https://github.com/SalesforceCommerceCloud/plugin_slas) team to align on the cookie expiry time

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`


# How to Test-Drive This PR

- `npm start` in /packages/pwa
- open localhost:3000
- verify the tokens are now stored as cookies
- verify guest users have cookie `cc-nx-g` and registered users have cookie `cc-nx`
- after registered user logout, the `cc-nx` tokens should be gone and a new `cc-nx-g` token should be created
- verify an OCAPI session is created upon login, a cookie `dwsid` should exist

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
